### ISSUE TITLE
"Structured field value" isn't well defined

### ISSUE THREAD

**Issue by jyasskin** [whatwg/fetch#1672]
https://fetch.spec.whatwg.org/#concept-header-list-set-structured-header takes a "[structured field value](https://httpwg.org/specs/rfc8941.html#rfc.section.2)" as an argument, but RFC8941 never defines that term. In particular, it's not clear that the argument needs to specify which of the [Structured Data Types](https://httpwg.org/specs/rfc8941.html#types) is being passed. This will often be guessable from the context, but Integers and Decimals might get mixed up in some cases, as could String, Tokens, and Byte Sequences.

**Comment by annevk** [comment]
It defines it just before the example at the end of https://httpwg.org/specs/rfc8941.html#rfc.section.2.

Perhaps Fetch should have a section though with some kind of mapping between those and Infra types.

@domenic might have some thoughts about this given his work around this in HTML.

**Comment by mnot** [comment]
Parameters do complicate things. Also, it'd be good to distinguish between `Token` and `String` if it's going to be (re-)serialised. And don't forget we're adding `Date` and perhaps a separate `Display String` that takes non-ASCII content.

@annevk if you remember we talked about this some at the last HTTP Workshop; the data model we eventually ended up looked something like:

~~~ json
{
  "@type": "string",
  "@val": "foo",
  "a": 1,
  "b": 2
}
~~~

... where `a` and `b` are Parameters. You could also have convenience methods for various things here, of course, but in this approach you'd have to make sure their names didn't confict with potential parameters. The other way would be to put parameters into a separate namespace; e.g.,

~~~
{
  "type": "string",
  "val": "foo",
  "params": {
    "a": 1,
    "b": 2
  }
}
~~~

**Comment by annevk** [comment]
Yeah, though note that this is for the "internal" API between specifications.

**Comment by mnot** [comment]
Ah.

**Comment by mnot** [comment]
Note that the [revision](https://httpwg.org/http-extensions/draft-ietf-httpbis-sfbis.html) is adding Dates and Display Strings (i.e., unicode).
